File: src/renderer/src/services/ipc/sftp/file.ts
import { backendService } from '../backend'

export const readFile = async (sessionId: string, path: string, binary?: boolean): Promise<string> => {
  const data = await backendService.request<{ content: string }>(
    {
      type: 'sftp:readfile',
      session_id: sessionId,
      data: { path, binary: binary ?? false }
    },
    'sftp:readfile',
    10000,
  )

  return data.content
}

export const writeFile = (sessionId: string, path: string, content: string): Promise<void> => {
  return backendService.request<void>(
    {
      type: 'sftp:writefile',
      session_id: sessionId,
      data: { path, content }
    },
    'sftp:writefile',
    10000,
  )
}

File: backend/internal/ipc/handlers/sftp/file.go
package sftp

import (
	"encoding/json"
	"fmt"
	"freessh-backend/internal/ipc/handlers"
	"freessh-backend/internal/models"
)

func (h *Handler) handleReadFile(msg *models.IPCMessage, writer handlers.ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.ReadFileRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse read file request: %w", err)
	}

	content, err := h.manager.ReadFile(msg.SessionID, req.Path, req.Binary)
	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPReadFile,
		SessionID: msg.SessionID,
		Data:      models.ReadFileResponse{Content: content, Path: req.Path},
	})
}

func (h *Handler) handleWriteFile(msg *models.IPCMessage, writer handlers.ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.WriteFileRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse write file request: %w", err)
	}

	if err := h.manager.WriteFile(msg.SessionID, req.Path, req.Content); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPWriteFile,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "saved", "path": req.Path},
	})
}
